mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
   }
   return ret;
 }
+
+enum DimType { Upper = 0, Lower = 1 };
+
+struct TransformAttrArgs {
+  rock::TransformType type;
+  std::pair<SmallVector<StringRef>, SmallVector<StringRef>> preservedNames;
+  std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> preservedDims;
+  SmallVector<int64_t> params;
+};
+
+template <DimType Type>
+SmallVector<uint32_t>
+getPreservedIndices(rock::TransformAttr tr,
+                    const SetVector<int64_t> &globalRemoveIndicesSet) {
+  SmallVector<uint32_t> preservedIndices;
+  auto upperDims =
+      Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  for (auto dim : upperDims) {
+    if (!globalRemoveIndicesSet.contains(dim)) {
+      preservedIndices.push_back(dim);
+    }
+  }
+  return preservedIndices;
+}
+
+template <DimType Type>
+void propagatePreservedNames(rock::TransformAttr tr, TransformAttrArgs &args) {
+  const auto &preservedDims = std::get<Type>(args.preservedDims);
+  auto &preseredNames = std::get<Type>(args.preservedNames);
+  auto names = Type == DimType::Upper ? tr.getUpperNames() : tr.getLowerNames();
+  auto dims = Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  assert(names.size() == dims.size());
+  for (auto [idx, dim] : llvm::enumerate(dims)) {
+    if (llvm::is_contained(preservedDims, dim)) {
+      preseredNames.push_back(names[idx]);
```

Comment:
Is this empty  before ? 
If so this does not feel like a propogation but rather a population of TransformAttrArgs struct.

Reply:
Can you add an assert make sure its empty ?

Reply:
I think we can do the following:

```cpp
template <DimType Type>
void propagatePreservedNames(rock::TransformAttr tr, TransformAttrArgs &args) {
  const auto &preservedDims = std::get<Type>(args.preservedDims);
  auto names = Type == DimType::Upper ? tr.getUpperNames() : tr.getLowerNames();
  auto dims = Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
  assert(names.size() == dims.size());

  SmallVector<StringRef> preservedNames = {};
  for (auto [idx, dim] : llvm::enumerate(dims)) {
    if (llvm::is_contained(preservedDims, dim)) {
      preservedNames.push_back(names[idx]);
    }
  }
  std::get<Type>(args.preservedNames) = std::move(preservedNames);
}
```

It becomes clear that `preservedNames` is empty before entering the loop

Reply:
Renamed `propagatePreservedNames` -> `populatePreservedNames`